python/ray/data/_internal/execution/operators/union_operator.py

```diff
@@ -0,0 +1,77 @@
+from typing import Callable, List, Optional, Tuple
+
+from ray.data._internal.execution.interfaces import PhysicalOperator, RefBundle
+from ray.data._internal.execution.operators.base_physical_operator import NAryOperator
+from ray.data._internal.stats import StatsDict
+from ray.data.block import BlockMetadata
+
+
+class UnionOperator(NAryOperator):
+    """An operator that combines blocks from input operators.
+
+    The order of the blocks from the input operators is preserved,
+    as is the relative ordering between the input operators.
```

Comment:
I don't think we should guarantee ordering unless the preserve order option is set.

Reply:
Prior to streaming executor, we do preserve the order. https://github.com/ray-project/ray/blob/a1e703e461b0f5b5e09e306c3e72697310ade001/python/ray/data/dataset.py#L1560
Also, union in SQL DBs preserves the order as well. 
It'd be weird if we don't preserve the order.

Reply:
Yes, but we made the decision to not preserve order by default in streaming mode. It's weird if Union behaves differently, regardless of legacy behavior.

Reply:
I see.

Reply:
I have re-implemented using a fully streaming approach as discussed in https://github.com/ray-project/ray/pull/36242#discussion_r1224681646, which takes from the first available input dependency (and therefore does not attempt to preserve order). Once the input dependency is exhausted, I have the output buffer take from the input buffer which has the greatest number of blocks.